packages/store/addon/-private/identifiers/cache.ts

```diff
@@ -0,0 +1,426 @@
+import {
+  StableRecordIdentifier,
+  IS_IDENTIFIER,
+  DEBUG_CLIENT_ORIGINATED,
+  DEBUG_IDENTIFIER_BUCKET,
+  GenerationMethod,
+  UpdateMethod,
+  ForgetMethod,
+  ResetMethod,
+} from '../ts-interfaces/identifier';
+import coerceId from '../system/coerce-id';
+import uuidv4 from './utils/uuid-v4';
+import normalizeModelName from '../system/normalize-model-name';
+import isStableIdentifier from './is-stable-identifier';
+import isNonEmptyString from '../utils/is-non-empty-string';
+
+type IdentifierMap = Dict<string, StableRecordIdentifier>;
+type TypeMap = Dict<string, KeyOptions>;
+
+interface KeyOptions {
+  lid: IdentifierMap;
+  id: IdentifierMap;
+  _allIdentifiers: StableRecordIdentifier[];
+}
+
+let configuredForgetMethod: ForgetMethod;
+let configuredGenerationMethod: GenerationMethod;
+let configuredResetMethod: ResetMethod;
+let configuredUpdateMethod: UpdateMethod;
+
+export function setIdentifierGenerationMethod(method: GenerationMethod): void {
+  configuredGenerationMethod = method;
+}
+
+export function setIdentifierUpdateMethod(method: UpdateMethod): void {
+  configuredUpdateMethod = method;
+}
+
+export function setIdentifierForgetMethod(method: ForgetMethod): void {
+  configuredForgetMethod = method;
+}
+
+export function setIdentifierResetMethod(method: ResetMethod): void {
+  configuredResetMethod = method;
+}
+
+function defaultGenerationMethod(data: ResourceIdentifierObject, bucket: string): string {
+  if (isNonEmptyString(data.lid)) {
+    return data.lid;
+  }
+  let { type, id } = data;
+  if (isNonEmptyString(id)) {
+    return `@ember-data:lid-${normalizeModelName(type)}-${id}`;
+  }
+  return uuidv4();
+}
+
+const IdentifierCaches = new WeakMap<object, IdentifierCache>();
+
+export function identifierCacheFor(store: object): IdentifierCache {
+  let cache = IdentifierCaches.get(store);
+
+  if (cache === undefined) {
+    cache = new IdentifierCache();
+    IdentifierCaches.set(store, cache);
+  }
+
+  return cache;
+}
+
+function defaultEmptyCallback(...args: any[]): any {}
+
+let DEBUG_MAP;
+if (DEBUG) {
+  DEBUG_MAP = new WeakMap<StableRecordIdentifier, StableRecordIdentifier>();
+}
+
+export class IdentifierCache {
+  // Typescript still leaks private properties in the final
+  // compiled class, so we may want to move these from _underscore
+  // to a WeakMap to avoid leaking
+  private _cache = {
+    lids: Object.create(null) as IdentifierMap,
+    types: Object.create(null) as TypeMap,
+  };
+  private _generate: GenerationMethod;
+  private _update: UpdateMethod;
+  private _forget: ForgetMethod;
+  private _reset: ResetMethod;
+
+  constructor() {
+    // we cache the user configuredGenerationMethod at init because it must
+    // be configured prior and is not allowed to be changed
+    this._generate = configuredGenerationMethod || defaultGenerationMethod;
+    this._update = configuredUpdateMethod || defaultEmptyCallback;
```

Comment:
Why do we have to set an empty callback?

Reply:
it's that or doing a `typeof fn === 'function'` check every time